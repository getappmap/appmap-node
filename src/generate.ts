import assert from "assert";
import type { ESTree } from "meriyah";

export function call_(
  callee: ESTree.Expression,
  ...args: ESTree.Expression[]
): ESTree.CallExpression {
  return {
    type: "CallExpression",
    callee,
    arguments: args,
  };
}

export function member(...ids: ESTree.Expression[]): ESTree.MemberExpression {
  assert(ids.length > 1);
  let result = [...ids];
  while (result.length > 1) {
    const [object, property, ...rest] = result;
    result = [
      { type: "MemberExpression", object, property, computed: property.type !== "Identifier" },
      ...rest,
    ];
  }
  assert(result[0].type === "MemberExpression");
  return result[0];
}

export type LiteralValue = ESTree.Literal["value"];

export function literal(value: LiteralValue): ESTree.Literal {
  return { type: "Literal", value };
}

export function identifier(id: string): ESTree.Identifier {
  return {
    type: "Identifier",
    name: id,
  };
}

export function ret(argument: ESTree.Expression | null = null): ESTree.ReturnStatement {
  return {
    type: "ReturnStatement",
    argument,
  };
}

export const args = identifier("arguments");
export const this_: ESTree.ThisExpression = { type: "ThisExpression" };
