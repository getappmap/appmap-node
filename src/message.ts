import console from "node:console";
import { inspect } from "node:util";

import chalk from "chalk";

export function info(message: unknown, ...args: unknown[]): void {
  console.info(prefix(message), ...args);
}

export function warn(message: unknown, ...args: unknown[]): void {
  console.warn(prefix(message), ...args);
}

function prefix(message: unknown): string {
  return chalk.magentaBright`(Î›) ` + (typeof message === "string" ? message : inspect(message));
}
