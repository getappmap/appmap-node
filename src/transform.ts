import assert from "node:assert";
import { writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { debuglog } from "node:util";
import { isNativeError } from "node:util/types";

import { generate } from "astring";
import { fromSource as getSourceMap, removeComments } from "convert-source-map";
import { parse, type ESTree } from "meriyah";

import applySourceMap from "./applySourceMap";
import * as instrument from "./hooks/instrument";
import * as jest from "./hooks/jest";
import * as mocha from "./hooks/mocha";
import { warn } from "./message";

const treeDebug = debuglog("appmap-tree");

export interface Hook {
  shouldInstrument(url: URL): boolean;
  transform(program: ESTree.Program): ESTree.Program;
}

const defaultHooks: Hook[] = [mocha, jest, instrument];

export function findHook(url: URL, hooks = defaultHooks) {
  return hooks.find((h) => h.shouldInstrument(url));
}

export default function transform(code: string, url: URL, hooks = defaultHooks): string {
  const hook = findHook(url, hooks);
  if (!hook) return code;

  try {
    const tree = parse(code, { source: url.toString(), next: true, loc: true, module: true });
    const sourcemap: unknown = getSourceMap(fixSourceMap(url, code))?.sourcemap;
    if (sourcemap) applySourceMap(tree, sourcemap);
    const xformed = hook.transform(tree);
    if (treeDebug.enabled) dumpTree(xformed, url);
    return generate(xformed);
  } catch (e) {
    assert(isNativeError(e));
    warn("Error transforming source at %s: %s", url, e.message);
    return code;
  }
}

function dumpTree(xformed: ESTree.Program, url: URL) {
  if (url.protocol !== "file:") return;
  const path = fileURLToPath(url) + ".appmap-tree.json";
  writeFileSync(path, JSON.stringify(xformed, null, 2));
  treeDebug("wrote transformed tree to %s", path);
}

// HACK: In node 18/16, when using --loader node-ts/esm
// sourcemap gets inserted twice to the typescript file. We remove the second one
// which reflects transpiled files map, instead of the original typescript file map.
function fixSourceMap(url: URL, code: string): string {
  const [major] = process.versions.node.split(".").map(Number);
  if (major >= 20) return code;

  if (![".ts", ".mts", ".tsx"].some((e) => url.pathname.toLowerCase().endsWith(e))) return code;

  let removed = removeComments(code);
  if (removed.indexOf("//# sourceMappingURL") > -1) {
    // Correct one remains, it needs to start in new line.
    removed = removed.replace("//# sourceMappingURL", "\n//# sourceMappingURL");
    return removed;
  }
  return code;
}
