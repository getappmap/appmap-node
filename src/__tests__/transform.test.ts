/* eslint-disable @typescript-eslint/unbound-method */
import { ESTree } from "meriyah";
import transform, { Hook } from "../transform";
import { literal } from "../generate";
import { warn } from "../message";

describe(transform, () => {
  it("uses hooks to transform the code", () => {
    const testHooks: Hook[] = [new TestHook(/foo/), new TestHook(/bar/)];
    const transforms = testHooks.map((t) => jest.spyOn(t, "transform"));

    expect(transform("'hello'", new URL("file:///foo"), testHooks)).toBe("'hello';\n\"/foo/\";\n");
    expect(transforms[0]).toBeCalled();
    expect(transforms[1]).not.toBeCalled();
    transforms[0].mockReset();

    expect(transform("'hello'", new URL("file:///bar"), testHooks)).toBe("'hello';\n\"/bar/\";\n");
    expect(transforms[0]).not.toBeCalled();
    expect(transforms[1]).toBeCalled();
    transforms[1].mockReset();

    expect(transform("'hello'", new URL("file:///other"), testHooks)).toBe("'hello'");
    expect(transforms[0]).not.toBeCalled();
    expect(transforms[1]).not.toBeCalled();
  });

  it("returns original code on error", () => {
    expect(
      transform("'hello'", new URL("file:///foo"), [
        {
          shouldInstrument() {
            return true;
          },
          transform() {
            throw new Error("test error");
          },
        },
      ]),
    ).toBe("'hello'");

    expect(jest.mocked(warn).mock.calls).toMatchSnapshot();
  });

  it("transforms unicode correctly", () => {
    const input = `const a = "\\u{1D306}";\n`;
    // Ensure that "\u{1D306}" is not converted to "tÌ†".
    expect(transform(input, new URL("file:///foo"), [new IdentityHook()])).toBe(input);
  });
});

class IdentityHook implements Hook {
  shouldInstrument(): boolean {
    return true;
  }
  transform(program: ESTree.Program): ESTree.Program {
    return program;
  }
}

class TestHook implements Hook {
  constructor(public pattern: RegExp) {}

  shouldInstrument(url: URL): boolean {
    return this.pattern.test(url.toString());
  }

  transform(program: ESTree.Program): ESTree.Program {
    return {
      ...program,
      body: [
        ...program.body,
        {
          type: "ExpressionStatement",
          expression: literal(this.pattern.toString()),
        },
      ],
    };
  }
}

jest.mock("../message");
