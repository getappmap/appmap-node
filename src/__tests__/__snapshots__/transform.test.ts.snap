// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform returns original code on error 1`] = `
[
  [
    "Error transforming source at %s: %s",
    "file:///foo",
    "test error",
  ],
]
`;
